Builders/MacOSBuilder.cs

```diff
@@ -24,25 +29,42 @@ public MacOSBuilder(string version, string? arch)
                 Logger.Error($"Invalid Architecture: {arch}");
 
             RuntimeIdentifier = $"{os_name}-{arch}";
+
+            stagingTarget = Path.Combine(Program.StagingPath, app_dir);
+            publishTarget = Path.Combine(stagingTarget, "Contents", "MacOS");
         }
 
         protected override string TargetFramework => "net8.0";
         protected override string RuntimeIdentifier { get; }
 
         public override Uploader CreateUploader()
         {
-            string extraArgs = $" --icon=\"{IconPath}\"";
+            string extraArgs = $" --signEntitlements=\"{Path.Combine(Environment.CurrentDirectory, "osu.entitlements")}\""
+                               + $" --noInst";
 
             if (!string.IsNullOrEmpty(Program.AppleCodeSignCertName))
                 extraArgs += $" --signAppIdentity=\"{Program.AppleCodeSignCertName}\"";
             if (!string.IsNullOrEmpty(Program.AppleInstallSignCertName))
                 extraArgs += $" --signInstallIdentity=\"{Program.AppleInstallSignCertName}\"";
             if (!string.IsNullOrEmpty(Program.AppleNotaryProfileName))
                 extraArgs += $" --notaryProfile=\"{Program.AppleNotaryProfileName}\"";
+            if (!string.IsNullOrEmpty(Program.AppleKeyChainPath))
+                extraArgs += $" --keychain=\"{Program.AppleKeyChainPath}\"";
 
-            return new VelopackUploader(app_name, os_name, RuntimeIdentifier, RuntimeIdentifier, extraArgs);
+            return new VelopackUploader(app_name, os_name, RuntimeIdentifier, RuntimeIdentifier, extraArgs: extraArgs, stagingPath: stagingTarget);
         }
 
-        public override void Build() => RunDotnetPublish();
+        public override void Build()
+        {
+            if (Directory.Exists(stagingTarget))
+                Directory.Delete(stagingTarget, true);
+
+            Program.RunCommand("cp", $"-r \"{Path.Combine(Program.TemplatesPath, app_dir)}\" \"{stagingTarget}\"");
+
+            RunDotnetPublish(outputDir: publishTarget);
+
+            // without touching the app bundle itself, changes to file associations / icons / etc. will be cached at a macOS level and not updated.
+            Program.RunCommand("touch", $"\"{stagingTarget}\" {Program.StagingPath}", false);
```

Comment:
Is there a reason why the `cp` and `touch` commands can't be implemented using .NET? Besides the fact that these commands are only usable on MacOS, I'd like to avoid using commands unless necessary.

If this can be engineered to use .NET filesystem methods, I'd prefer that.

Reply:
Maybe, but I've just copied the original code here. Not particularly interested in changing that unless there's a world-breaking event.

As in #1, you can't fully use this thing on a non-MacOS system anyway.